app/wasm/test/custom_msg_test.go

```diff
@@ -0,0 +1,469 @@
+	StarPool  uint64
+	AtomPool  uint64
+	RegenPool uint64
+}
+
+func TestSwapMsg(t *testing.T) {
+	// table tests with this setup
+	cases := map[string]struct {
+		msg        func(BaseState) *wasmbindings.SwapMsg
+		expectErr  bool
+		initFunds  sdk.Coin
+		finalFunds []sdk.Coin
+	}{
+		"exact in: simple swap works": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "ustar",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(5000000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 13000000),
+			finalFunds: []sdk.Coin{
+				sdk.NewInt64Coin("uosmo", 1000000),
+				sdk.NewInt64Coin("ustar", 120000000),
+			},
+		},
+		"exact in: price too low": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "ustar",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(555000000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 13000000),
+			expectErr: true,
+		},
+		"exact in: not enough funds to swap": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "ustar",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(5000000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 7000000),
+			expectErr: true,
+		},
+		"exact in: invalidPool": {
+			msg: func(state BaseState) *wasmbindings.SwapMsg {
+				return &wasmbindings.SwapMsg{
+					First: wasmbindings.Swap{
+						PoolId:   state.StarPool,
+						DenomIn:  "uosmo",
+						DenomOut: "uatom",
+					},
+					// Note: you must use empty array, not nil, for valid Rust JSON
+					Route: []wasmbindings.Step{},
+					Amount: wasmbindings.SwapAmountWithLimit{
+						ExactIn: &wasmbindings.ExactIn{
+							Input:     sdk.NewInt(12000000),
+							MinOutput: sdk.NewInt(100000),
+						},
+					},
+				}
+			},
+			initFunds: sdk.NewInt64Coin("uosmo", 13000000),
+			expectErr: true,
+		},
+
+		// FIXME: this panics in GAMM module !?! hits a known TODO
```

Comment:
I hit this two times with ExactOut. It seems that if you trade from A -> B where 1 A is worth > 2 B then the ExactOut swaps will all fail.

When I reverse it such that the output is the more valuable currency, then ExactOut passes.